### PR TITLE
Create cid

### PR DESCRIPTION
**Pull request by RobStallion**
Create functions that create same CIDs as IPFS. Function takes a string, map or struct.

~Currently uses ex_cid~

#7 #11 

Have (temporarily) renamed the module to ex_cid. See #22 for more info on this. 

Adds tests and doc tests.

### CODE REVIEW

File: lib/ex_cid.ex [hunk 1]
```diff
@@ -1,42 +1,27 @@
-defmodule DwylCid do
-  defstruct [:a]
+defmodule ExCid do
```
Comment (nelsonic) [inline lib/ex_cid.ex]: https://hex.pm/packages/excid name available on hex.pm ✅ 
image

File: mix.exs [hunk 2]
```diff
@@ -21,7 +21,9 @@ defmodule Rid.MixProject do
   # Run "mix help deps" to learn about dependencies.
   defp deps do
     [
-      {:ex_multihash, "~> 2.0"}
+      {:ex_multihash, "~> 2.0"},
+      {:jason, "~> 1.1"},
+      {:cid, git: "git@github.com:SimonLab/ex-cid.git"}
```
Comment (nelsonic) [inline mix.exs]: This project should not have a dependency on `ex-cid`.

**Comment by RobStallion** [comment]
@nelsonic I have updated this now 👍

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+language: elixir
```
Comment (RobStallion) [inline .travis.yml:1]: 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 

Good addition @nelsonic

**Review by nelsonic — CHANGES_REQUESTED** [review]
@RobStallion this is looking _good_ so far and I _could_ just _merge_ it. 
But we _both_ know it's not "ready" to be used by another project.
 
Please add tests for more "complex" data e.g: 
```elixir
animal = %{
  name: "Rex",
  type: "dog",
  legs: 4
}
```
or
```elixir
task = %{
  uid: 1,
  title: "Get into the top 1% of people in my chosen field of expertise https://git.io/top",
  done: false
}
```
Then Document it like your _life_ depends on it. 📝 
For this you will need to approach it with a ["Beginners Mind"](https://en.wikipedia.org/wiki/Shoshin) ...
Look at the code & docs with a fresh pair of eyes and ask yourself:
would a _complete_ beginner understand _everything_ about this?
If the answer is no (_which, hint: at this point it should be..._), add in-line comments
until you feel that the code/examples could be understood by someone on their _first_ day of learning elixir.

File: lib/cid.ex [hunk 4]
```diff
@@ -1,38 +1,87 @@
 defmodule Cid do
   @moduledoc """
-  Returns a SHA512 transformed to Base64, remove ambiguous chars then sub-string
+  Provides a way for a user to turn a String, Map or Struct into a CID that
+  is identical to one what will be returned from IPFS if the same data is
+  added.
+
+  Currently only produces a default v1 CID.
+  Currently only uses the "raw" codec
```
Comment (Danwhy) [inline lib/cid.ex:8]: Maybe add a little more info on what a v1 CID or "raw" codec is. Just a link would probably be fine, even if it's to elsewhere in this repo.

File: lib/cid.ex [hunk 5]
```diff
@@ -1,38 +1,87 @@
 defmodule Cid do
   @moduledoc """
-  Returns a SHA512 transformed to Base64, remove ambiguous chars then sub-string
+  Provides a way for a user to turn a String, Map or Struct into a CID that
+  is identical to one what will be returned from IPFS if the same data is
```
Comment (Danwhy) [inline lib/cid.ex]: A link to an explanation of what IPFS is might be good here

File: lib/cid.ex [hunk 6]
```diff
@@ -1,38 +1,87 @@
...
+  defp create_multihash(_), do: {:error, "invalid data type"}
+
+  # if an error is passed in return error message
+  defp create_cid({:error, msg}), do: msg
+
+  # takes a multihash and retuns a CID
+  defp create_cid(multihash) when is_binary(multihash) do
+    multihash
+    |> create_cid_suffix()
+    |> B58.encode58()
```
Comment (Danwhy) [inline lib/cid.ex:89]: Could do with a brief explanation of B58

File: lib/cid.ex [hunk 7]
```diff
@@ -1,38 +1,87 @@
...
+    |> add_multibase_prefix()
+  end
+
+  # takes a multihash and returns the suffix
+  # currently version is hardcoded to 1
+  # and multicodec-packed-content-type is hardcoded to "raw" ("U" == <<85>>)
+  # <version><multicodec-packed-content-type><multihash>
+  defp create_cid_suffix(multihash), do: <<1>> <> "U" <> multihash
```
Comment (Danwhy) [inline lib/cid.ex:102]: A small comment on the binary syntax might be useful to people unfamiliar with it

File: lib/cid.ex [hunk 8]
```diff
@@ -1,38 +1,87 @@
...
-    input |> stringify_map_values |> make
+  def cid(value) do
+    value
+    |> create_multihash()
+    |> create_cid()
+  end
+
+  # if create_multihash is called with a struct, the struct is converted into a
+  # map and then create_multihash is called again
+  defp create_multihash(%_{} = struct) do
```
Comment (Danwhy) [inline lib/cid.ex:50]: A brief explanation of how this pattern matching works would be useful for beginners

File: lib/cid.ex [hunk 9]
```diff
@@ -1,38 +1,87 @@
...
+    |> Jason.encode!()
+    |> create_multihash()
+  end
+
+  # if create_multihash is called with a string, the string has a new line added
+  # to the end (as that's what IPFS appears to be doing based on tests), then
+  # the string is converted into a multihash
+  defp create_multihash(str) when is_binary(str) do
+    str = add_new_line(str)
+    digest = :crypto.hash(:sha256, str)
```
Comment (Danwhy) [inline lib/cid.ex:69]: Explaining in the comments that this is an erlang function would be good for beginners

File: lib/cid.ex (hunk 6 above)
Comment (RobStallion) [reply lib/cid.ex:89]: Just added this in [one of the latest commits](https://github.com/dwyl/cid/pull/21/commits/8a21996c8b71957e14fa8beeac6df69c3ba71551). Thanks for pointing that out

**Comment by RobStallion** [comment]
@Danwhy Thanks for all the points on the documentation updates that can be done here. I'll get on those right away. All really good points thanks 👍

File: lib/cid.ex (hunk 5 above)
Comment (nelsonic) [reply lib/cid.ex]: Please link to it in the `README.md` in https://github.com/dwyl/learn-ipfs i.e. "What is IPFS?" and if we don't already have the section, we should add it ASAP see: https://github.com/dwyl/learn-ipfs/issues/1#issuecomment-458522501

**Comment by RobStallion** [comment]
@nelsonic @Danwhy Have updated the pr based on your feedback.

**Review by nelsonic — CHANGES_REQUESTED** [review]
@RobStallion looking better. 👍 
[_Again_](https://youtu.be/fhrNgXJ__n8?t=217) ... 🥋

File: lib/cid.ex [hunk 10]
```diff
@@ -33,6 +33,7 @@ defmodule Cid do
     iex> Cid.cid([1,2,3,"four"])
     "invalid data type"
   """
+  @spec cid(String.t | map() | struct()) :: String.t
```
Comment (nelsonic) [inline lib/cid.ex:36]: 👍

**Comment by nelsonic** [comment]
@RobStallion & @Danwhy _good_ work on adding property-based-testing. 🎉 
_Showcase_ your work by adding a Travis-CI and Codecov Badge ... 😉

**Comment by RobStallion** [comment]
@nelsonic Will do.

**Review by nelsonic — APPROVED** [review]
@RobStallion looking _much_ better. Thanks! 🎉 
Can't _wait_ to apply this to our https://github.com/dwyl/phoenix-ecto-append-only-log-example 😉

**Pull request merged**